### PR TITLE
Add a parallel mode to the stream listener

### PR DESCRIPTION
**Pull request by meln1k**
# Description

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://some.engineering/code-of-conduct).

### CODE REVIEW

File: fixcloudutils/redis/event_stream.py [hunk 1]
```diff
@@ -271,6 +305,12 @@ async def read_all() -> None:
         await self.__outdated_messages_task.start()
 
     async def stop(self) -> Any:
+        async def stop_task(task: Task[Any]) -> None:
+            task.cancel()
+            with suppress(asyncio.CancelledError):
+                await task
```
Comment (aquamatthias) [inline fixcloudutils/redis/event_stream.py]:
```suggestion
        await stop_running_task(task)
```

File: fixcloudutils/redis/event_stream.py [hunk 2]
```diff
@@ -165,8 +171,13 @@ async def _listen(self) -> None:
...
                 self.__readpos = ">"
-
-                await self._handle_stream_messages(messages)
+                if self.parallelism:
+                    await self._handle_stream_messages_parallel(messages, self.parallelism)
+                else:
+                    await self._handle_stream_messages(messages)
+            except RuntimeError as r:
+                if len(r.args) > 0 and r.args[0] == "no running event loop":
+                    raise r
```
Comment (aquamatthias) [inline fixcloudutils/redis/event_stream.py]: This catches all other runtime errors and probably differs from what you wanted.

File: fixcloudutils/redis/event_stream.py [hunk 3]
```diff
@@ -185,6 +196,28 @@ async def _handle_stream_messages(self, messages: List[Any]) -> None:
...
                 await self.redis.xack(self.stream, self.group, *ids)
 
+    async def _handle_stream_messages_parallel(self, messages: List[Any], max_parallelesm: int) -> None:
+        """
+        Handle messages in parallel in unordered fasion. The number of parallel tasks is limited by max_parallelism.
+        """
+
+        async def handle_and_ack(msg: Any, message_id: StreamIdT):
+            await self._handle_single_message(msg)
+            await self.redis.xack(self.stream, self.group, message_id)
```
Comment (aquamatthias) [inline fixcloudutils/redis/event_stream.py:210]: Not sure how redis handles acknowledgment of messages arrived later, while previous messages are not done yet or even failing?

Comment (meln1k) [reply fixcloudutils/redis/event_stream.py:210]: If I understand the documentation correctly, redis should handle this well. I also wrote a small test, which tests the out of order acknowledgement and it works.

File: fixcloudutils/redis/event_stream.py [hunk 4]
```diff
@@ -185,6 +195,28 @@ async def _handle_stream_messages(self, messages: List[Any]) -> None:
                 # acknowledge all processed messages
                 await self.redis.xack(self.stream, self.group, *ids)
 
+    async def _handle_stream_messages_parallel(self, messages: List[Any], max_parallelesm: int) -> None:
+        """
+        Handle messages in parallel in unordered fasion. The number of parallel tasks is limited by max_parallelism.
```
Comment (aquamatthias) [inline fixcloudutils/redis/event_stream.py]:
```suggestion
        Handle messages in parallel in an unordered fashion. The number of parallel tasks is limited by max_parallelism.
```

File: fixcloudutils/redis/event_stream.py (hunk 1 above)
Comment (aquamatthias) [reply fixcloudutils/redis/event_stream.py]: See: fixcloudutils.asyncio.stop_running_task

File: fixcloudutils/redis/event_stream.py [hunk 5]
```diff
@@ -185,6 +195,28 @@ async def _handle_stream_messages(self, messages: List[Any]) -> None:
...
+            await self.redis.xack(self.stream, self.group, message_id)
+
+        def task_done_callback(task: Task[Any]) -> None:
+            self._ongoing_tasks.discard(task)
+
+        for stream, stream_messages in messages:
+            log.debug(f"Handle {len(stream_messages)} messages from stream.")
+            for uid, data in stream_messages:
+                if len(self._ongoing_tasks) >= max_parallelesm:  # queue is full, wait for a slot to be freed
+                    await asyncio.wait(self._ongoing_tasks, return_when=asyncio.FIRST_COMPLETED)
```
Comment (aquamatthias) [inline fixcloudutils/redis/event_stream.py]: Could lead to possibly more than max_parrallel tasks. Consider using a semaphore.

**Review by aquamatthias — APPROVED** [review]

**Pull request merged**